src/core/qgsogrproxytextcodec.cpp

```diff
@@ -0,0 +1,98 @@
+/***************************************************************************
+                             qgsogrproxytextcodec.cpp
+                             -------------
+    begin                : June 2020
+    copyright            : (C) 2020 Nyall Dawson
+    email                : nyall dot dawson at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsogrproxytextcodec.h"
+#include "qgslogger.h"
+#include <cpl_string.h>
+#include <mutex>
+
+QgsOgrProxyTextCodec::QgsOgrProxyTextCodec( const QByteArray &name )
+  : mName( name )
+{
+
+}
+
+QString QgsOgrProxyTextCodec::convertToUnicode( const char *chars, int, ConverterState * ) const
+{
+  if ( !chars )
+    return QString();
+
+  char *res = CPLRecode( chars, mName.constData(), CPL_ENC_UTF8 );
+  if ( !res )
+  {
+    QgsDebugMsg( "convertToUnicode failed" );
+    return QString();
+  }
+
+  const QString result = QString::fromUtf8( res );
+  CPLFree( res );
+  return result;
+}
+
+QByteArray QgsOgrProxyTextCodec::convertFromUnicode( const QChar *unicode, int length, ConverterState * ) const
+{
+  if ( !unicode )
+    return QByteArray();
+
+  const QString src = QString( unicode, length );
+  char *res = CPLRecode( src.toLocal8Bit().constData(), CPL_ENC_UTF8, mName.constData() );
```

Comment:
shouldn't that rather be something like (untested) ?
```suggestion
  char *res = CPLRecode( src.toUtf8().constData(), CPL_ENC_UTF8, mName.constData() );
```